hub/hub_client.go

```diff
@@ -0,0 +1,123 @@
+package hub
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+	"web-chat-go/user"
+
+	"github.com/gorilla/websocket"
+)
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+}
+
+type Client struct {
+	User *user.User
+	Hub  *Hub
+	Conn *websocket.Conn
+	Send chan []byte
+}
+
+type ClientTCP struct {
+	User *user.User
+	Conn net.Conn
+}
+
+// clientClose will unregister the given Client and close it's connection
+// to the web application.
+func (c *Client) clientClose() {
+	c.Hub.unregister <- c
+	c.Conn.Close()
+}
+
+// readMessages defers the closure of the Client and enables an
+// implementation of a Reader logic that will read each message sent from the
+// web application through the WebSocket. The read message will be broadcasted
+// to the entire Hub's collection of clients.
+func (c *Client) readMessages() {
+	defer c.clientClose()
+
+	for {
+		// Read message from browser
+		_, msg, err := c.Conn.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(
+				err, websocket.CloseGoingAway,
+				websocket.CloseAbnormalClosure,
+			) {
+				fmt.Println("The websocket closed unexpectedly.")
+				break
+			} else if websocket.IsCloseError(
+				err, websocket.CloseGoingAway,
+			) {
+				fmt.Printf("User %v has logged out.\n", c.User.UserName)
+				break
+			}
+			panic(err)
+		}
+
+		tNow := []byte(fmt.Sprint(time.Now().Format("2006-01-02 15:04:05"), "|"))
```

Comment:
you should have a message object instead of formatting it here.  otherwise, what will happen when a message is sent with pipes in it?